site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/repository/HapiFhirRepository.java

```diff
@@ -37,92 +37,138 @@
         return responseElement.getIdElement().getIdPart();
     }
 
     /**
-     * @param researchStudy the researchStudy to save.
-     * @return ResearchStudy
+     * @param organization the organization to save.
+     * @return id of the saved organization
      */
-    public String saveResearchStudy(ResearchStudy researchStudy) {
-        // Log the request
-
-        logger.info(FhirRepo.SAVE_REQUEST.message(),
-                    fhirContextWrapper.prettyPrint(researchStudy));
+    public String saveOrganization(final Organization organization) {
+        return saveResource(organization);
+    }
 
-        Bundle responseBundle;
+    /**
+     * Exact search on name
+     *
+     * @param name of the organization to search.
+     * @return org with that name or none() if not found
+     */
+    @Override
+    public Optional<Organization> findOrganizationByName(final String name) {
         try {
-            responseBundle = fhirContextWrapper.executeTransaction(fhirUri,
-                    bundle(researchStudy, RESEARCHSTUDY_ENTITY_NAME));
+            return fhirContextWrapper
+                    .search(fhirUri, Organization.class)
+                    .where(Organization.NAME.matchesExactly().value(name))
+                    .returnBundle(Bundle.class)
+                    .execute()
+                    .getEntry()
+                    .stream()
+                    .findFirst()
+                    .map(Bundle.BundleEntryComponent::getResource)
+                    .filter(r -> r.getResourceType().equals(ResourceType.Organization))
+                    .map(Organization.class::cast);
         } catch (BaseServerResponseException e) {
-            logger.warn(FhirRepo.UPDATE_ERROR.message(), e);
-            throw new RestException(e.getMessage(), e);
+            throw new RestException(String.format(Repositorys.SEARCH_ERROR.message(), e.getMessage()), e);
         }
-        IBaseResource responseElement = extractResponseResource(responseBundle);
-
-        // Log the response
-        logger.info(FhirRepo.SAVE_RESPONSE.message(),
-                    fhirContextWrapper.prettyPrint(responseElement));
+    }
 
-        return responseElement.getIdElement().getIdPart();
+    /**
+     * @param researchStudy the researchStudy to save.
+     * @return ResearchStudy
+     */
+    public String saveResearchStudy(final ResearchStudy researchStudy) {
+        return saveResource(researchStudy);
     }
 
     private IBaseResource extractResponseResource(Bundle bundle) throws RestException {
-        var resp = fhirContextWrapper.toListOfResources(bundle);
-        
+        final var resp = fhirContextWrapper.toListOfResources(bundle);
         if (resp.size() != 1) {
-            logger.info(FhirRepo.BAD_RESPONSE_SIZE.message(), resp.size());
-
+            logger.info(Repositorys.BAD_RESPONSE_SIZE.message(), resp.size());
```

Comment:
would this logger be better in the exception? Saves code duplication.

Reply:
Done, just throws doesn't log now (will be logged at top level anyway).